nova/virt/vmwareapi/vm_util.py

```diff
@@ -2053,3 +2056,40 @@ def rename_vm(session, vm_ref, instance):
 def is_vim_instance(o, vim_type_name):
     return isinstance(o, sudsobject.Factory.subclass(vim_type_name,
                                                      sudsobject.Object))
+
+
+def service_locator_name_password(client_factory, username, password):
+    o = client_factory.create('ns0:ServiceLocatorNamePassword')
+    o.username = username
+    o.password = password
+    return o
+
+
+def _get_sha1_ssl_thumbprint(url):
+    """Returns the sha-1 thumbprint of the ssl cert of the vcenter
+    """
+    parsed_url = six.moves.urllib.parse.urlparse(url)
+    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    sock.settimeout(1)
+    wrappedSocket = ssl.wrap_socket(sock)
+    thumb = None
+    wrappedSocket.connect((parsed_url.hostname, parsed_url.port or 443))
+
+    der_cert_bin = wrappedSocket.getpeercert(True)
+
+    thumb = hashlib.sha1(der_cert_bin).hexdigest().upper()
+    t = iter(thumb)
+    return ':'.join(a + b for a, b in zip(t, t))
```

Comment:
Aren't there a loads of errors the functions called in here could raise?

Reply:
I wasn't aware of this pattern ... create an iterator out of a list and then zip the iterator to merge the two objects right after another into a single entry? I think I would have used `zip(t[::2], t[1::2])`
```
>>> a = [1, 2]
>>> b = iter(a)
>>> zip(b, b)
[(1, 2)]
>>> a = [1, 2, 3, 4]
>>> b = iter(a)
>>> zip(b, b)
[(1, 2), (3, 4)]
>>> zip(a[::2], a[1::2])
[(1, 2), (3, 4)]
```
Funny idea.